cilium: switch to mq/fq and add bandwidth manager

borkmann opened the pull request:
One item from our data path optimization project is to switch to BBR model with underlying FQ (for single queue devs) or MQ + FQ leafs (for multi queue devs) in tc qdisc layer. This also allows us to support K8s `kubernetes.io/egress-bandwidth` per Pod annotations in a fully lockless way via EDT and BPF which has been implemented in this PR as well. Verified via netperf that the egress-bandwidth matches the one the Pod was configured with. The agent option is a simple `--enable-bandwidth-manager` switch.

borkmann committed:
cilium: add bandwidth manager

Base layer which implements setup of BBR + {MQ/FQ, FQ} as well as
EDT based rate-limiting in BPF. Agent code implements map setup and
handling of egress bandwidth label for Pods.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: disable timer migration

Avoid spinlock contention on resetting TCP retransmit timer
for large servers. Not disabling timer_migration is known to
cause issues on larger servers with >= 64 CPUs.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Link: https://lore.kernel.org/netdev/CANn89iKS6fas9O74U5w1wb+8DN==fXRKQ8nzq0tkT_VOXRtYBQ@mail.gmail.com

borkmann committed:
cilium, api: extend openapi to dump bandwidth manager status

Dump info on whether it is enabled and if so which devices are managed.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, status: add bandwidth manager status

Add status info message about bandwidth manager. Example output with agent
flag --enable-bandwidth-manager=true and supported kernel:

  # ./cilium/cilium status
  [...]
  BandwidthManager:       EDT with BPF   [eno1]
  [...]

Otherwise:

  # ./cilium/cilium status
  [...]
  BandwidthManager:       Disabled
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, bpf: add metric counter for EDT drop horizon

Add a counter so we can inspect rate-limiting drops due to too large
departure time windows.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, docs: add gsg for bandwidth manager

Add a simple GSG guide for the bandwidth manager along with a deployment
example users can verify.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,199 @@
+	return &EdtInfo{v.Bps, v.TimeLast, v.TimeHorizonDrop, v.Pad}
+}
+
+var ThrottleMap = bpf.NewMap(
```
This should live in `pkg/maps/bandwidth`.

joestringer commented on pkg/endpoint/types.go:
```diff
@@ -0,0 +1,30 @@
+	IdentityLabels labels.Labels
+	InfoLabels     labels.Labels
+	Annotations    map[string]string
+}
```
nit: I wonder if we should combine this with `pkg/endpoint/regeneration` for a new dedicated endpoint types package, to avoid pulling in the rest of the `pkg/endpoint` dependencies in packages like `pkg/k8s`

borkmann replied:
I didn't go the route via cgroups, the merged set is completely independent of it.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -879,6 +880,10 @@ int from_netdev(struct __ctx_buff *ctx)
+	/* Traffic from the host ns going through cilium_host device must
+	 * not be subject to EDT rate-limiting.
+	 */
+	edt_set_aggregate(ctx, 0);
```
I'm not sure I understand why this is needed/what it achieves. Packets from the host namespace going to the wire will be directly routed to the native device without going through the host device, so they will not go through `from-host`.

borkmann replied:
Not necessarily, for example, consider the case when we have tunneling enabled, this would go via cilium_host -> cilium_vxlan -> phys dev. In this case we want to make sure that EDT aggregate is 0 and not subject to rate-limiting, too. (The rate-limiting implemented here also works with tunneling.)

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -233,11 +233,10 @@ func initKubeProxyReplacementOptions() (strict bool) {
+func detectHostDevices(strict bool) {
 	detectNodePortDevs := len(option.Config.Devices) == 0 &&
-		(option.Config.EnableNodePort || option.Config.EnableHostFirewall)
+		(option.Config.EnableNodePort || option.Config.EnableHostFirewall || option.Config.EnableBandwidthManager)
```
We still have a couple of bugs for when devices are set but our kube-proxy replacement is disabled (e.g., #12141 and #12205). This is partly why the host firewall GSG sets `kubeProxyReplacement=strict`. I don't think there's much we can do here (apart from fixing the bugs), but mentioning it here so we keep in mind users may run into these bugs just because they enabled this new option while using kube-proxy.

EDIT: I just saw that this is enabled by default for new deployments through the Helm chart. Maybe worth settings the native devices only if the kernel actually supports `bpf_skb_ecn_set_ce`?

borkmann replied:
Ah okay, is anyone currently fixing these bugs?

borkmann replied:
I've added the prerequisite probing for disabling device detection if BW manager is the only one that would need it.

borkmann replied:
The two mentioned bugs we simply need to get fixed.

pchaigno replied:
> Ah okay, is anyone currently fixing these bugs?

I don't think anyone is at the moment. They'll be next on my host-fw list once I've fixed the bigger issues with the host firewall in kube-proxy mode.

borkmann replied:
Okay, sgtm, thanks!

pchaigno commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,144 @@
+			log.WithError(err).WithField("device", device).Warn("Link does not exist")
+			continue
+		}
+		// We strictly want to avoid an down/up cycle on the device at
```
Nit:
```suggestion
		// We strictly want to avoid a down/up cycle on the device at
```

pchaigno commented on bpf/lib/edt.h:
```diff
@@ -0,0 +1,76 @@
+		return CTX_ACT_OK;
+	}
+	if (t_next - now >= info->t_horizon_drop)
+		return CTX_ACT_DROP;
```
I'm not sure I understand `t_horizon_drop`. I understand what it does (i.e., drops packets that should be sent too far in the future), but I don't understand why we need this (don't lower layers take care of that already?) or how we should define its value. Also, it seems to be constant, so maybe it doesn't need to be part of the map value?

borkmann replied:
I've added a comment to the code explaining what it does and why it's needed.

pchaigno commented on Documentation/gettingstarted/bandwidth-manager.rst:
```diff
@@ -0,0 +1,120 @@
+CNI plugin.
+
+Cilium's bandwidth manager supports the ``kubernetes.io/egress-bandwidth`` Pod
+annotation which is enforced on egress at the native host networking devices.
```
Why don't we apply this rate limiting sooner (at lxc devices)? I couldn't find anything in the K8s documentation that requested either implementation but may have missed something.

borkmann replied:
veth devices are `noqueue` devices by default, it would be broken to somehow enforce queueing on tc ingress in the host ns. It's possible via ifb, but completely broken architecture. The only point that makes sense is the tc egress on the phys dev. Also updated the doc wrt ingress shaping.

pchaigno replied:
Ah, I didn't know that, but yeah, it makes sense :slightly_smiling_face:

borkmann committed:
cilium: add bpf bandwidth command for map introspection

Add a tool for introspecting current bandwidth settings from the BPF map:

  # ./cilium/cilium bpf bandwidth list
  IDENTITY   EGRESS BANDWIDTH (BitsPerSec)
  491        125M

This can easily correlated with `cilium endpoint list`.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno reviewed (commented):
LGTM overall, just have a few questions and concerns below. Not super familiar with queuing and congestion control in Linux, so I may missing some things.

aanm commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,147 @@
+	"github.com/cilium/cilium/pkg/maps/bwmap"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/sirupsen/logrus"
```
new line between 23 and 24

aanm commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+const (
+	Subsystem = "bandwidth-manager"
```
can be private `s/Subsystem/subsystem`

aanm commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,147 @@
+		return
+	}
+
+	log.Infof("Setting up BPF bandwidth manager")
```
`s/Infof/Info`

aanm commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,147 @@
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.SysParamName:  s.name,
+				logfields.SysParamValue: s.val,
+			}).Warning("Failed to set sysctl needed by BPF bandwidth manager. Disabling the feature.")
```
Alright we disabled the feature but why aren't we reverting the sysctl values that we previously successfully wrote?

borkmann replied:
These should actually be Fatal(). Seems in the GH action smoke-test that in the Cilium container procfs does not have these available whereas in normal deployment these are fine to set. So I'll add a probe function and disable the BW manager if we are in such env.

aanm commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,147 @@
+
+	if _, err := bwmap.ThrottleMap.OpenOrCreate(); err != nil {
+		log.WithError(err).Fatal("Failed to access ThrottleMap")
+	}
```
We could do this operation before we write our `sysctl` values.

aanm commented on pkg/endpoint/events.go:
```diff
@@ -179,3 +184,64 @@ func (ev *EndpointPolicyVisibilityEvent) Handle(res chan interface{}) {
+	bpsOld := "inf"
+	bpsNew := "inf"
+	if e.bps != 0 {
+		bpsOld = fmt.Sprintf("%v", e.bps)
```
`bpsOld = strconv.FormatUint(e.bps, 10)`

aanm commented on pkg/endpoint/events.go:
```diff
@@ -179,3 +184,64 @@ func (ev *EndpointPolicyVisibilityEvent) Handle(res chan interface{}) {
+		bpsOld = fmt.Sprintf("%v", e.bps)
+	}
+	if bps != 0 {
+		bpsNew = fmt.Sprintf("%v", bps)
```
`bpsNew = strconv.FormatUint(bps, 10)`

aanm commented on pkg/maps/bwmap/bwmap.go:
```diff
@@ -0,0 +1,76 @@
+	// Flow aggregate is per Pod, so same size as Endpoint map.
+	MapSize = lxcmap.MaxEntries
+
+	DefaultTimeHorizon = 2000 * 1000 * 1000
```
Please add a comment to what's this magic number?

aanm commented on pkg/maps/bwmap/bwmap.go:
```diff
@@ -0,0 +1,76 @@
+	DefaultTimeHorizon = 2000 * 1000 * 1000
+)
+
+type EdtId struct {
```
add comment to what `edt` means

borkmann replied:
Added to doc.go

aanm commented on api/v1/models/bandwidth_manager.go:
```diff
@@ -0,0 +1,51 @@
+// Code generated by go-swagger; DO NOT EDIT.
```
This commit should be squashed with the previous one.

aanm commented on Documentation/gettingstarted/bandwidth-manager.rst:
```diff
@@ -0,0 +1,132 @@
+annotation which is enforced on egress at the native host networking devices.
+
+The ``kubernetes.io/ingress-bandwidth`` annotation is not supported and also not
+recommended to use.
```
>also not recommended to use.

please add the answer to the "Why?" in here.

aanm commented on Documentation/gettingstarted/bandwidth-manager.rst:
```diff
@@ -0,0 +1,132 @@
+###########
+
+    * Bandwidth enforcement currently does not work when traffic needs to go through
+      an L7 proxy in the host namespace.
```
```
    * Bandwidth enforcement currently does not currently work with L7 Cilium Network Policies. In case they select the pod, either at ingress or egress, the bandwidth enforcement will be disabled.
```

borkmann replied:
Thanks for the suggestion, I refined this a bit.. only relevant for egress policies.

aanm reviewed (changes requested):
Also, misses e2e tests.

brb commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -233,11 +233,10 @@ func initKubeProxyReplacementOptions() (strict bool) {
-// (if needed).
-func detectDevicesForNodePortAndHostFirewall(strict bool) {
+// detectHostDevices tries to detect bpf_host devices (if needed).
+func detectHostDevices(strict bool) {
```
nit: s/detectHostDevices/detectNativeDevices/?

brb commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,159 @@
+		{"net.ipv4.tcp_max_syn_backlog", "4096"},
+	}
+	for _, s := range baseSettings {
+		log.Infof("Setting sysctl %s=%s", s.name, s.val)
```
nit: we should be consistent with other log statements and thus, use structured logging (`WithFields`).

borkmann replied:
Ok, will do. We'll need to make it consistent then with the other sysctls that are being set before init.sh, so I'll add `WithFields` to both then.

brb commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,159 @@
+		val  string
+	}
+	baseSettings := []setting{
+		{"net.core.netdev_max_backlog", "1000"},
```
Maybe worth adding a comment or two about the setting values?

brb commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,159 @@
+			log.WithError(err).WithField("device", device).Warn("Link does not exist")
+			continue
+		}
+		// We strictly want to avoid a down/up cycle on the device at
```
Q: this means that a previously attached bpf_host won't get temporarily detached?

borkmann replied:
No, this is the actual tc 'egress' qdisc which won't affect sch_clsact egress.

brb commented on pkg/datapath/loader/base.go:
```diff
@@ -291,7 +291,7 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 
 	clockSource := []string{"ktime", "jiffies"}
-	log.Infof("Setting up base BPF datapath (BPF %s instruction set, %s clock source)",
+	log.Infof("Setting up BPF datapath (BPF %s instruction set, %s clock source)",
```
nit: Same comment re structured logging.

brb commented on pkg/datapath/maps/map.go:
```diff
@@ -165,6 +165,10 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
 	}
 
+	if !option.Config.EnableBandwidthManager {
+		maps = append(maps, "cilium_throttle")
```
s/"cilium_throttle"/bwmap.MapName/?

borkmann replied:
Not a single map in this list uses the package name. Probably begs for some refactoring.

brb commented on pkg/k8s/watchers/pod.go:
```diff
@@ -224,7 +225,9 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
 	newAnno := newK8sPod.ObjectMeta.Annotations
-	annotationsChanged := !k8s.AnnotationsEqual([]string{annotation.ProxyVisibility}, oldAnno, newAnno)
+	annoChangedProxy := !k8s.AnnotationsEqual([]string{annotation.ProxyVisibility}, oldAnno, newAnno)
+	annoChangedBandwidth := !k8s.AnnotationsEqual([]string{bandwidth.EgressBandwidth}, oldAnno, newAnno)
```
I think we should log a warning if a user sets an ingress bandwidth saying that it's going to be ignored.

borkmann replied:
Makes sense, I'll add that.

brb commented on pkg/endpoint/events.go:
```diff
@@ -179,3 +184,64 @@ func (ev *EndpointPolicyVisibilityEvent) Handle(res chan interface{}) {
+	if bandwidthEgress != "" {
+		bps, err = bandwidth.GetBytesPerSec(bandwidthEgress)
+		if err == nil {
+			err = (&bwmap.ThrottleBPFMap{}).Update(e.ID, bps)
```
nit: Instead of creating a struct each time invoking `Update/Delete`, you could just create an instance of the struct in the `bwmap` pkg, and all then invoke the methods for it.

borkmann replied:
This is an empty struct, I've actually copied this from ipmasq map code. ;)

borkmann replied:
Removed.

brb replied:
Yeah, in ipmasq we use the struct to be able to mock the ipmasq map in unit tests.

brb commented on bpf/lib/edt.h:
```diff
@@ -0,0 +1,80 @@
+	ctx->queue_mapping = aggregate;
+}
+
+static __always_inline __u32 edt_get_aggregate(struct __ctx_buff *ctx)
```
nit: `edt_{set,get}_aggregate_id()`?

borkmann replied:
I think okay as-is, no strong opinion though.

brb reviewed (commented):
LGTM, just a few nits.

borkmann committed:
cilium: use structured logging for setting sysctl params

Use structured logging (WithFields) for sysctl settings:

  level=info msg="Setting up BPF bandwidth manager" subsys=bandwidth-manager
  level=info msg="Setting sysctl" subsys=bandwidth-manager sysParamName=net.core.netdev_max_backlog sysParamValue=1000
  level=info msg="Setting sysctl" subsys=bandwidth-manager sysParamName=net.core.somaxconn sysParamValue=4096
  level=info msg="Setting sysctl" subsys=bandwidth-manager sysParamName=net.core.default_qdisc sysParamValue=fq
  level=info msg="Setting sysctl" subsys=bandwidth-manager sysParamName=net.ipv4.tcp_congestion_control sysParamValue=bbr
  level=info msg="Setting sysctl" subsys=bandwidth-manager sysParamName=net.ipv4.tcp_max_syn_backlog sysParamValue=4096
  level=info msg="Setting qdisc to fq" device=eno1 subsys=bandwidth-manager
  level=info msg="Setting up BPF datapath" bpfClockSource=ktime bpfInsnSet=v2 subsys=datapath-loader
  level=info msg="Setting sysctl" subsys=datapath-loader sysParamName=net.core.bpf_jit_enable sysParamValue=1
  level=info msg="Setting sysctl" subsys=datapath-loader sysParamName=net.ipv4.conf.all.rp_filter sysParamValue=0
  level=info msg="Setting sysctl" subsys=datapath-loader sysParamName=kernel.unprivileged_bpf_disabled sysParamValue=1
  level=info msg="Setting sysctl" subsys=datapath-loader sysParamName=kernel.timer_migration sysParamValue=0
  level=info msg="Setting sysctl" subsys=datapath-loader sysParamName=net.ipv6.conf.all.disable_ipv6 sysParamValue=0

Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: warn on unsupported bandwidth pod annotations

Throw a warning in particular when users try to configure an ingress bandwidth
limitation to make it obvious that we do not support this. Also add one for
egress when bw manager is disabled.

Example when kubernetes.io/ingress-bandwidth is set:

  [...]
  level=warning msg="Endpoint has kubernetes.io/ingress-bandwidth annotation which is unsupported. This annotation is ignored." annotations="map[kubernetes.io/egress-bandwidth:125M kubernetes.io/ingress-bandwidth:125M]" k8sPodName=default/netperf-6d6cd89c86-6xtgh subsys=daemon
  [...]

Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

aanm commented on test/k8sT/Bandwidth.go:
```diff
@@ -0,0 +1,136 @@
+			cmd := helpers.Netperf(targetIP, helpers.TCP_MAERTS, "-l 30 -I 99,5")
+			for _, pod := range pods {
+				By("Making netperf request from %s pod to pod with IP %s", pod, targetIP)
+				res := kubectl.ExecPodCmd(helpers.DefaultNamespace, pod, cmd)
```
I don't understand one thing, this is running netperf between 2 pods but where do we actually check that there was a rate limit being applied?

borkmann replied:
Yeah, I've extracted it locally so far, still need to implement in the test. Currently swamped with Plumbers though. Will rebase, get the below CI failures passing & implement the range check (thinking +/-5Mbit range check).

kkourt commented on pkg/endpoint/events.go:
```diff
@@ -179,3 +184,65 @@ func (ev *EndpointPolicyVisibilityEvent) Handle(res chan interface{}) {
+			err = bwmap.Update(e.ID, bps)
+		}
+	} else {
+		err = bwmap.Delete(e.ID)
```
AFAICT, this (when there is no configuration for an endpoint) is the only point  where we delete ids from the map. I'm wondering if we need to do the same when the user restarts cilium with bandwidth manager disabled, but I'm not sure if the old entries persist or not during a restart.

pchaigno replied:
@kkourt The map is deleted if a user restarts with the bandwidth manager disabled (cf. the change in `pkg/datapath/maps/map.go`).

kkourt commented on bpf/lib/edt.h:
```diff
@@ -0,0 +1,80 @@
+
+	/* We need to reset queue mapping here such that new mapping will
+	 * be performed based on skb hash. See netdev_pick_tx().
+	 */
```
Just to make sure that my reading is correct,  are we using `->queue_mapping` to store the endpoint ID because this field will not be used between the time we set it and reset it here?

Is there a reason we are naming this aggregate?

borkmann replied:
Correct, queue mapping is retained all the way as opposed to others. Naming is aggregate since multiple sockets are aggregated into the same EDT class for rate-limiting.

kkourt commented on Documentation/gettingstarted/bandwidth-manager.rst:
```diff
@@ -0,0 +1,144 @@
+Bandwidth Manager (beta)
+************************
+
+This guide explains how to configure Cilium's bandwidth manager to in order to
```
```suggestion
This guide explains how to configure Cilium's bandwidth manager to
```

joestringer reviewed (commented):
Apparently I had this comment queued up but never posted it. Anyway I'll post it now in case it's still relevant.
